### PR TITLE
Naheed & Dionisia's Ride Share Edges

### PR DESCRIPTION
**Pull request by larachan15**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer | Instructor Feedback
:------------- | :------------- | ---
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | We realized that sequence mattered for certain tests to pass. For instance in the trip_dispatcher file, we created the load_drivers definition under load_users because it felt logical to group users and drivers in the same area. However, then we had to move @trips below @drivers in the initialize definition because otherwise we had errors in our test code.
Describe any examples of composition that you encountered in this project, if any | I think we used composition when we made the module RideShare. By using the module, we were able to call functions in other classes. | By composition we mean a _has-a_ or _has-many_ relationship. Examples in this project include "a TripDispatcher _has-many_ Drivers, Passengers and Trips", "a Driver _has-many_ Trips", "a Trip _has-a_ Driver".
Describe the relationship between `User` and `Driver`| User is the superclass of Driver. Driver inherited the attributes of User, plus had additional one such as driven_trips and status.
Describe a _nominal test_ that you wrote for this assignment. |  We wrote a test to check if the total costs for all rides per user was calculated correctly.
Describe an _edge case test_ that you wrote for this assignment | We wrote a test to check the average rating returned zero if the driven has not driven any trips.
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | We started to understand how all the files related to each other better and how to call the methods in each of the classes. Also, we started to understand how the driver file related to users and trips files.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We both like giving and receiving upfront feedback from each other. Also, we both had different levels of understanding in this project that helped us each figure out different portions, and then we were able to explain our understanding to the other.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+          revenue = (trip.cost - 1.65) * 0.8
+          income += revenue
+        end
+      end
+      return income.round(2)
+    end
+
+    def net_expenditures
+      return super - total_revenue
+    end
```
Comment (droberts-sea) [inline lib/driver.rb:79]: Good use of `super` here!

File: lib/user.rb [hunk 2]
```diff
@@ -16,5 +17,29 @@ def initialize(input)
     def add_trip(trip)
       @trips << trip
     end
+
+    def net_expenditures
+      cost_array = []
+      @trips.each do |trip|
+        if trip.cost == nil
+          raise ArgumentError, "Trip is in progress, no cost"
```
Comment (droberts-sea) [inline lib/user.rb:25]: I don't know that raising an `ArgumentError` is the right course of action here - you could imagine a user wanting to look up their net expenditures while on a trip, for example. Instead I would probably omit incomplete trips:

```ruby
trips.each do |trip|
  if trip.cost.nil?
    next
  end
  # ... add to the total ...
end
```

Or even better, you could write a helper method that returns a list of complete trips:

```ruby
def completed_trips
  return @trips.reject { |t| t.end_time.nil? }
end

def net_expenditures
  completed_trips.each do |trip|
    # ... same logic as before ...
  end
end
```

File: lib/driver.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+      @vin = input[:vin]
+      @driven_trips = input[:driven_trips].nil? ? [] : input[:driven_trips]
+      @status = input[:status] == nil ? :AVAILABLE : input[:status]
+    end
+
+    def average_rating
+      total_ratings = 0
+      @driven_trips.each do |trip|
+        total_ratings += trip.rating
```
Comment (droberts-sea) [inline lib/driver.rb:25]: All these driver methods will fail for incomplete trips!

File: specs/user_spec.rb [hunk 4]
```diff
@@ -59,4 +60,39 @@
...
+                                 cost: 40,
+                                 rating: 5)
+      @user.add_trip(trip)
+
+    end
+    
+    it 'calculates total cost of all rides per user' do
+      @user.trips.each do |trip|
+        expect(@user.net_expenditures).must_equal 80
+      end
```
Comment (droberts-sea) [inline specs/user_spec.rb:89]: For both of these methods, what if the user has no trips? What if they have an incomplete trip?

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -1,104 +1,156 @@
...
+      it "returns a float within range of 1.0 to 5.0" do
+        average = @driver.average_rating
+        expect(average).must_be :>=, 1.0
+        expect(average).must_be :<=, 5.0
+      end
+
+      it "returns zero if no driven trips" do
+        driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV",
+                                       vin: "1C9EVBRM0YBC564DZ")
+        expect(driver.average_rating).must_equal 0
```
Comment (droberts-sea) [inline specs/driver_spec.rb:86]: Good job catching this edge case.

File: lib/trip_dispatcher.rb [hunk 6]
```diff
@@ -25,52 +29,132 @@ def load_users(filename)
...
+
+      def request_trip(user_id)
+        chosen_driver = assign_driver(user_id)
+        passenger = find_passenger(user_id)
 
-    private
+        valid_trip_id = [1..1000].select
+        @trips.each do |trip|
+          if trip.id == valid_trip_id
+            raise ArgumentError, "ID already exists"
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:112]: I don't think this code will do what you want. `[1..1000].select` returns a thing called an enumerator, which is basically a loop with no block. Aside from not being a reasonable value for an ID, this will never be equal to your integer trip ID, so the argument error will never be raised.

Moreover, even if you had said `.sample` instead of `.select`, raising an error because you happened to generate a bad random number is not a great behavior. Instead, you should retry selecting a number, or even better choose the number without randomness:

```ruby
highest_current_id = @trips.map { |t| t.id }.max
new_trip_id = highest_current_id + 1
```

File: lib/trip_dispatcher.rb [hunk 7]
```diff
@@ -25,52 +29,132 @@ def load_users(filename)
...
+          end
+        end
 
-    def check_id(id)
-      raise ArgumentError, "ID cannot be blank or less than zero. (got #{id})" if id.nil? || id <= 0
+        trip = RideShare::Trip.new(id: valid_trip_id, driver: chosen_driver, passenger: passenger, start_time: Time.now, end_time: nil, cost: nil, rating: nil)
+
+        chosen_driver.add_driven_trip(trip)
+        passenger.add_trip(trip)
+        @trips << trip
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:120]: You don't mark the driver as unavailable!

File: lib/trip_dispatcher.rb [hunk 8]
```diff
@@ -25,52 +29,132 @@ def load_users(filename)
...
+        return trip
+      end
+
+      def assign_driver(passenger_id)
+        # iterates through the drivers to select and return available drivers and drivers who are not driving themselves
+        available_drivers = @drivers.select do |driver|
+          if driver.status == :AVAILABLE && driver.id != passenger_id
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:127]: I like that you broke this logic out as a separate method! Good instincts.

File: lib/trip_dispatcher.rb [hunk 9]
```diff
@@ -25,52 +29,132 @@ def load_users(filename)
...
+          driver.driven_trips.empty?
+        end
+
+        # if there hasn't been a driver that hasn't given any trips, then this iterates through available drivers and selects the driver who's trip ended the longest time ago.
+        if chosen_driver.nil?
+          furthest_date = Time.now
+          available_drivers.each do |driver|
+            # This assumes that driven_trips are in chronological order
+            end_time = driver.driven_trips.last.end_time
+            if end_time < furthest_date
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:143]: That might be a big assumption.

File: lib/trip_dispatcher.rb [hunk 10]
```diff
@@ -25,52 +29,132 @@ def load_users(filename)
...
+
+        # if there are no available drivers
+        if chosen_driver.nil?
+          raise ArgumentError, "No drivers available"
+        end
+
+        # if there is an available driver, changes their status
+        chosen_driver.status = :UNAVAILABLE
+        return chosen_driver
+      end
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:158]: Aha, here's where you set the status. I would probably not do this yet - what if something goes wrong in trying to create the trip? Then the driver would be unavailable but not actually have an incomplete trip.

File: lib/trip_dispatcher.rb [hunk 11]
```diff
@@ -25,52 +29,132 @@ def load_users(filename)
...
+        available_drivers = @drivers.select do |driver|
+          if driver.status == :AVAILABLE && driver.id != passenger_id
+            driver
+          end
+        end
+
+        # iterates through the available_drivers to see if there is a driver that has not given any trips.
+        chosen_driver = available_drivers.find do |driver|
+          driver.driven_trips.empty?
+        end
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:135]: Good use of enumerables throughout this method!

File: lib/driver.rb [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+    def total_revenue
+      revenue = 0.0
+      income = 0.0
+      @driven_trips.each do |trip|
+        if trip.cost == nil
+          raise ArgumentError, "Trip still in progress, no revenue"
+        else
+          revenue = (trip.cost - 1.65) * 0.8
+          income += revenue
```
Comment (droberts-sea) [inline lib/driver.rb:71]: This code would be a little more readable if the two magic numbers (`0.8` and `1.65`) were stored in constants, maybe something like `DRIVER_CUT` and `FEE`.

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes
Answer comprehension questions | yes - see feedback above
**Wave 1** |
Appropriate use of Ruby's `Date`| yes
`Trip` has a helper method to calculate duration| yes
`User` (passenger) has a method to calculate total cost of all trips| yes
Tests for wave 1 | missing some edge cases - see inline
**Wave 2** |
`Driver` inherits from `User` | yes
`Driver` has `add_driven_trip` method | yes
`Driver` has method to calculate average rating | yes
`Driver` has method to calculate net expenditures and it uses `super` | yes
`Driver` has a method to calculate total revenue | yes
Tests for wave 2 | yes
**Wave 3** |
`TripDispatcher` has a new method to create trips | yes
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | no (driver status is set directly)
Complex logic was correctly implemented | yes
Tests for `request_trip` | yes
Methods from wave 1 and 2 handle incomplete trips | sort of - see inline
Tests for wave 1 and 2 methods with incomplete trips | no
**Wave 4 (Optional)** |
`TripDispatcher` now assigns trips to either the newest driver (no trips), or the driver who has not driven in the longest time | yes
Appropriate helper methods were made to help with complex logic | yes
Tests for wave 4 | yes - great work!
**Overall** |

Great job overall! I'm very happy with the code you've written, and impressed by your work on the optionals.

At the risk of raining on the parade, I do want to call out that going back and working on previous projects is not always the best use of time, and I hope that you haven't let your Hotel implementations slip in order to put in extra effort here.

There are a few places where things could be cleaned up or where test coverage could be improved that I've tried to call out inline, but this is generally a very solid submission. Keep up the hard work!
